### PR TITLE
Java: Update Java JDK 17 models.

### PR DESCRIPTION
**Pull request by michaelnebel**
### TLDR
In this PR we re-generate the Java SDK 17 models using the *mixed* model generator. That is, we use the content based (field aware) model generation where applicable, otherwise we default to the old heuristic for detecting summaries and tainting entire objects. Some minor manual modelling was required after the re-generation. DCA shows
- No performance regressions.
- Removal of several false positives as they were flagged due to field conflation (with the non content based models).
- The new results based on
  - Content based summaries look accurate.
  - Non content based summaries look reasonable.
  - Manual summaries look accurate.

Even though it is hard to say (as there are lot of manual models for the Java SDK), it is my impression, that the *mixed* model generation improves the precision (quality) or our results.

### Details
The sections below contains more detailed information on test failures/updates and changes to alerts.

Some notes on the model and test updates:
- The update of the request forgery expected output is because the generated models now include `["java.net.http", "HttpRequest$Builder", True, "build", "()", "", "Argument[this]", "ReturnValue", "taint", "df-generated"]`, which propagates the taint to the `send` sink as well.
- `TopJdkApisTest.ql`: Due to the following neutrals, we no longer generate models for these callables (which is causing the change in the test output): 
  - `["java.util.concurrent", "Executor", "execute", "(Runnable)", "summary", "manual"]`
  - `["java.util.stream", "IntStream", "mapToObj", "(IntFunction)", "summary", "manual"]`
- `ApplicationModeEndpoints.ql`: This test reports reports model usages. The new set of models doesn't contain generated models for some of the method overloads for `java.io.PrintStream.println`. For these to be generated we should have a model for `String.valueOf(Object)`, which is explicitly added as a summary neutral (so this change in the generated models would also have happened even, if we didn't generate mixed models). In any case, the summary models for `PrintStream` are worthless as it is a wrapper class and there is not any models that retrieve any value from `this` (unless returning an entire `PrintStream` object). That is, IMO we can accept these changes.
- `LdapInjection.ql`. It turns out that the more precise models for `BasicAttributes` (which are correct) doesn't work properly in conjunction with the `manual` sink definition see below. Less precise models that works with the `search` sink has been added.
  - ["javax.naming.directory", "DirContext", True, "search", "", "", "Argument[0..1]", "ldap-injection", "manual"]
- The expected test output for the remaining tests have been accepted without any deeper investigation as it didn't effect reported results (only reported edges and nodes).

Some notes on DCA results:

Removed results:

- `java/exec-tainted-environment` on `apache/hive` was a false positive. Before we had the models seen below. However, the constructor *only* taints the `command` field, which is not used in the `environment` method and thus the combination of these models leads to false positive flow.
  - ["java.lang", "ProcessBuilder", False, "ProcessBuilder", "(String[])", "", "Argument[0].ArrayElement", "Argument[this]", "taint", "df-generated"]
  - ["java.lang", "ProcessBuilder", False, "environment", "()", "", "Argument[this]", "ReturnValue", "taint", "ai-manual"]
- `java/sensitive-log` on `apache/hive` were false positives. Same reason as for `java/exec-tainted-environment` on `apache-hive`.
- `java/sensitive-log` on `apache/geode` was a false positive. Before we had the models seen below. However, it is only `Argument[0]` that leads to tainting the returnvalue of `getType`.
  - ["javax.management", "Notification", True, "Notification", "(String,Object,long)", "", "Argument[0]", "Argument[this]", "taint", "df-generated"]
  - ["javax.management", "Notification", True, "Notification", "(String,Object,long)", "", "Argument[1]", "Argument[this]", "taint", "df-generated"]
  - ["javax.management", "Notification", True, "getType", "()", "", "Argument[this]", "ReturnValue", "taint", "df-generated"]

Added results
- `java/path-injection` on `WebGoat/WebGoat`: Due to the manual `ZipFile` models (this looks sound).
- `java/sensitive-log` on `apache/hadoop-common`: Due to non content based summaries for `javax.security.sasl.SaslServer` (this might be a it speculative).
- `java/sensitive-log` on `apache/solr`: Due to non content based summaries for `java.xml.stream.XMLStreamReader` (these look reasonable).
- `java/zipslip` on `apache/flink`: Due to content based summaries for `java.io.File` (these summaries looks good).
- `java/zipslip` on `apache/geode`: Same as above

### CODE REVIEW

**Review by owen-mc — APPROVED** [review]
It looks good. I don't totally follow your reasoning on `PrintStream`, but I'm happy to accept your decision.

I note that we make quite a lot of models for methods that override `Object.clone()`. There is special treatment for this [here](https://github.com/github/codeql/blob/4df9cd88ef170908f6b1ca3eae56913493c87404/java/ql/lib/semmle/code/java/dataflow/internal/DataFlowUtil.qll#L261) and [here](https://github.com/github/codeql/blob/4df9cd88ef170908f6b1ca3eae56913493c87404/java/ql/lib/semmle/code/java/dataflow/internal/DataFlowPrivate.qll#L424), so we don't need to model these methods. Is it worth putting an exclusion in the model generator for this? (We do have various manual models for it.)

**Review by owen-mc — APPROVED** [review]
Thanks for dealing with the clone methods

**Comment by michaelnebel** [comment]
> It looks good. I don't totally follow your reasoning on `PrintStream`, but I'm happy to accept your decision.
> 
> I note that we make quite a lot of models for methods that override `Object.clone()`. There is special treatment for this [here](https://github.com/github/codeql/blob/4df9cd88ef170908f6b1ca3eae56913493c87404/java/ql/lib/semmle/code/java/dataflow/internal/DataFlowUtil.qll#L261) and [here](https://github.com/github/codeql/blob/4df9cd88ef170908f6b1ca3eae56913493c87404/java/ql/lib/semmle/code/java/dataflow/internal/DataFlowPrivate.qll#L424), so we don't need to model these methods. Is it worth putting an exclusion in the model generator for this? (We do have various manual models for it.)

Thank you very much for the review and thank you for proactively asking on slack!!!

- Yes, it is indeed a good idea to exclude `clone` methods from the model generation. Thank you!
- After re-reading the description of why `PrintStream` models doesn't provide much value, I see that it is a bit convoluted: The `PrintStream` class is a wrapper like class which doesn't expose its internals after being created. Having summaries for such a class is of limited (or no value), so we can accept the *removal* (however, for *sink* modelling it would be a different story).

**Comment by michaelnebel** [comment]
@owen-mc : Needed to update one more test for the CI to pass. I intend to also run QA job before merging this PR (when I get back from PTO).

**Review by owen-mc — APPROVED** [review]

**Comment by aschackmull** [comment]
The detailed summary sounds good to me. I noticed one thing, though: `"java.util.stream", "IntStream", "mapToObj"` should not be neutral - it ought to have a manual value model from `"Argument[0].ReturnValue"` to `"ReturnValue.Element"`. The same goes for `mapToObj` on the other specialised stream classes, `DoubleStream` and `LongStream`.

**Comment by michaelnebel** [comment]
> The detailed summary sounds good to me. I noticed one thing, though: `"java.util.stream", "IntStream", "mapToObj"` should not be neutral - it ought to have a manual value model from `"Argument[0].ReturnValue"` to `"ReturnValue.Element"`. The same goes for `mapToObj` on the other specialised stream classes, `DoubleStream` and `LongStream`.

That makes sense. I think that should be considered potential follow up work.
Will start a QA run.

**Comment by michaelnebel** [comment]
The alert counts diff of a QA run of this branch (after a rebase) vs the latest nightly bundle shows
image

The QA run shows variations on a few repositories for path-injection, sensitive-log and log-injection, which are also the queries that produces the most results (by far). Since we don't see any particular overall decrease or increase in the alert count, I think we should accept these results without further investigation (the discrepancies were thoroughly investigated for the DCA execution and the conclusion was that we now produce more precise results with the new models).

@aschackmull @owen-mc : Do you have any objections? Otherwise, I think we should go ahead and merge the new models.
